### DISCUSSION TITLE
Making tables from files (and more)

### DISCUSSION THREAD

**Discussion by roaldarbol** [squidfunk/mkdocs-material#3315] [Ideas]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I want to suggest an idea and checked that ...

- [X] ... to my best knowledge, my idea wouldn't break something for other users
- [X] ... the documentation does not mention anything about my idea
- [X] ... there are no open or closed issues that are related to my idea

### Description

First of all, thanks for all the brilliant work you're putting into this project - it's my go-to for everything documentation now!
This is a multi-feature request, which I'd be happy to discuss further. I'm no JS wizard so I'm not sure if I can help much myself on that side.

I'm proposing enabling the following workflow:

- Query and read data file (e.g. filtering, choosing columns, and more)
- Render table in Markdown
- Render the Markdown table with .js (like now, but maybe with better options for customisation)

--- 

**Tables from files**
Being able to parse and query a data file (e.g. csv) to Markdown. There might be inspiration to get from the Obsidian [obsidian-csv-table plugin](https://github.com/coddingtonbear/obsidian-csv-table). They take the approach where you can read the file by: 

````
```csvtable
source: my_csv_file.csv
filter: population < 100000000
```
````

**More js options**
The above point is the important one! However, although being able to sort the table is good, it could easily be more. I'd think this should be relatively easy to implement (though I haven't tried and might be very wrong) e.g. with [DataTables](https://datatables.net/) which is open source and can be further customized with extensions.

### Use Cases

The data used in tables most often lives in data files (be it .csv, .txt, .xlsx etc.). And currently we have to type out the content of these files to render them as a table. If we'd be able to call a data file and render it directly, when something changes in the data, we wouldn't have to scour the pages to find the entries and replace them, they'd be changed automatically. I think this would be useful to almost everyone who has the need to make tables.

Personally, I'm currently working on an open hardware project where these have proved vital. We have a collection of parts in a file and want to showcase subsets of parts needed to make each build.

**Comment by facelessuser** [comment]
You could do this with a custom fence with SuperFences: https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#custom-fences.

The one thing to note is that when using SuperFence's custom fences, whatever is returned by the custom fence logic is stored away and replaced with a placeholder such that further Markdown parsing is not run on it. Placeholders are replaced with the actual content at the end.

You can do almost anything with them. We use it in the pymdown-extension documentation to create special blocks that spawn a new Markdown process such that we can display Markdown rendered with different options (specified with YAML frontmatter) than what the documents use: https://github.com/facelessuser/pymdown-extensions/blob/main/tools/pymdownx_md_render.py.  Then we just provide our blocks:

````markdown
```md-render
---
extensions:
- pymdownx.magiclink
- pymdownx.saneheaders
- markdown.extensions.attr_list

extension_configs:
  pymdownx.magiclink:
    repo_url_shortener: true
    repo_url_shorthand: true
    social_url_shorthand: true
    social_url_shortener: true
    user: facelessuser
    repo: pymdown-extensions
---
- Just paste links directly in the document like this: https://google.com.
- Or even an email address: fake.email@email.com{.magiclink-ignore}.
```
````

So, you could simply create your own that simply parses the code block as a CSV and return it as an HTML table.

**Comment by squidfunk** [comment]
Thanks for suggesting. Material for MkDocs is a theme, so what it does is mainly rendering. All Markdown processing is done by upstream projects, mainly Python Markdown and Python Markdown Extensions by @facelessuser. Preprocessing tabular data into Markdown tables is definitely Markdown processing, which is beyond the scope of this project. Material for MkDocs tries to make the most functional and beautiful website of the Markdown you're throwing at it, but that's basically it. Furthermore, preprocessing is theme-agnostic, since you're taking a format and converting it into Markdown, so it could probably be implemented as a Markdown extension.

I would love to improve data tables, but the Markdown table syntax doesn't allow for much flexibility. The DataTables project depends on jQuery, which we don't have as a dependency for several reasons. Do you know of any other good data tables libraries, ideally without any external dependencies? I could take a look at whether we can add native support, similar to [TableSort](https://squidfunk.github.io/mkdocs-material/reference/data-tables/#sortable-tables), which is also a third-party library.

**Comment by squidfunk** [comment]
Moving this into a discussion

**Comment by roaldarbol** [comment]
Thanks for the replies! @squidfunk that of course makes sense, I keep forgetting that its a theme. Maybe it would be worth considering adding this case to the docs (once we perhaps end up at a good solution). As for alternatives, I especially like [Simple DataTables](https://github.com/fiduswriter/Simple-DataTables) - it's also lightweight, has no external dependencies and seems to be be well-maintained so that might be worth considering. ([examples here](https://fiduswriter.github.io/Simple-DataTables/))

@facelessuser that seems a good approach. I've not previously worked with PyMdown or its extensions so I'm note quite sure how to go about this: Say I have a `.csv` file in `docs/data`. How do I parse that? Do I make a SuperFence with plain python code? And do I also explicitly need to parse that into HTML too? 
Ideally I'd hope to get to a solution like the one I mentioned above where, just as with the `mermaid` example you call the name just after the backticks, but I don't know if that's too intricate?

**Reply by facelessuser** [reply]
> Say I have a .csv file in docs/data. How do I parse that?

You could have your custom fence open it and read it in and then use a python CSV lib to parse it.

> How do I parse that?

There seems to be one built-in to Python, though I've never personally used it: https://docs.python.org/3/library/csv.html.

> And do I also explicitly need to parse that into HTML too?

Not sure what you mean, but you would return the HTML desired as a string.

So, you'd create the logic to read the content of the fence and turn it into HTML. Then register the custom fence in SuperFences.

Something like you showed above is definitely possible:

 ````
 ```csvtable
 source: my_csv_file.csv
 filter: population < 100000000
 ```
 ````
> Do I make a SuperFence with plain python code?

Yes, assuming I understand what you are asking. The example I gave was done with pure Python code: https://github.com/facelessuser/pymdown-extensions/blob/main/tools/pymdownx_md_render.py.

> Ideally I'd hope to get to a solution like the one I mentioned above where, just as with the mermaid example you call the name just after the backticks, but I don't know if that's too intricate?

Yeah, as I said, you do *exactly* what you proposed in the opening post. The fenced block would be replaced with whatever your custom fence logic returns.

I do not have the time to offer custom solutions, but if you get stuck on something, I can probably answer some specific questions.

**Reply by roaldarbol** [reply]
Sorry, I think there's just something plain and simple I'm missing. 

> Yes, assuming I understand what you are asking. The example I gave was done with pure Python 

I was not sure if the python code would go actually within the fences like in Rmarkdown or Jupyter Notebooks. I'm now beginning to get that I would write a script somewhere else called `csvtable.py` which is then called upon? As you say, that would achieve what I was asking for initially. And are there any best practices on where to keep the custom scripts?

> Not sure what you mean, but you would return the HTML desired as a string.

If I use python (`csv.reader()`) to import the data it still needs to be converted to a table. You mentioned that it should be converted directly to HTML as Markdown wouldn't be re-rendered if I understood it correctly. 

Sorry about the confusion, I think there's some implicit knowledge I'm missing here. Not asking for custom solutions, just need to understand how to achieve it. Thanks for the help, I really appreciate it!

**Reply by facelessuser** [reply]
> I was not sure if the python code would go actually within the fences like in Rmarkdown or Jupyter Notebooks. I'm now beginning to get that I would write a script somewhere else called `csvtable.py` which is then called upon? As you say, that would achieve what I was asking for initially.

Yes, external code is linked to a custom fence name which you can use just like `mermaid` is used.

> If I use python (`csv.reader()`) to import the data it still needs to be converted to a table. You mentioned that it should be converted directly to HTML as Markdown wouldn't be re-rendered if I understood it correctly.

Yes, you would convert it to HTML syntax. So it would be a string that of HTML: `"<table>...</table>"`. The caveat with using the SuperFences custom fence is that once you return that string, no **further** processing can be done on it. For things like what you are wanting, mermaid, etc., this is usually fine. In your case, you are likely just displaying data, not trying to provide more content for the Markdown parser to process.

In my example, which I linked, I get around the limitation of the content not further being processed by the Markdown parser by launching another Markdown process, but it is essentially sandboxed from the content outside of the fence, but this is perfect for what I wanted. Anyways, for most people, such a thing is not needed.

The point is that you can do anything you want to to the content in the fence, you just need to format the result exactly as you want it as that is what will end up in the final output.

**Reply by roaldarbol** [reply]
Ah, that makes sense, thanks! 
Last question: The generated HTML table, will that then be post-processed by JS e.g. the `tablesort` which mkdocs-material supports?

**Reply by roaldarbol** [reply]
From the [documentation](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#uml-diagram-example) my understanding is that the javascript would find the `<table>` blocks and render them.

**Reply by facelessuser** [reply]
JavaScript is run at browser runtime. That is long done after Python Markdown is through with it. The table will be available for `tablesort`, no problem.

**Comment by facelessuser** [comment]
> Not asking for custom solutions, just need to understand how to achieve it.

Understood. I like to set expectations up front as sometimes I am asked for more than I am willing to give 🙂.

**Reply by roaldarbol** [reply]
No worries, one has to set boundaries in this space! I found a good answer of yours to an issue on `pymdown-extensions` which clarifies things [here](https://github.com/facelessuser/pymdown-extensions/issues/928#issuecomment-624056104) - I think it provides a good minimal example for how to implement your own custom fencing which could be great documentation for us rookies. :-)

**Reply by facelessuser** [reply]
Yeah, hopefully, it is up to date, but at the very least, it hopefully gives you enough of an understanding of the basics.

**Comment by chrimaho** [comment]
I've found a really good solution for this.

You can read more on this blog: [The best MkDocs plugins and customizations](https://chrieke.medium.com/the-best-mkdocs-plugins-and-customizations-fc820eb19759).
Which references the plugin `mkdocs-table-reader-plugin`, as documented on [GitHub](https://github.com/timvink/mkdocs-table-reader-plugin) and their [Docs](https://timvink.github.io/mkdocs-table-reader-plugin/) site.

All you need to do is:
```sh
pip install mkdocs-table-reader-plugin
```
Then extend the `mkdocs.yml` like:
```yaml
plugins:
  - search
  - table-reader
```
And then add a CSV to your page like:
```md
{{ read_csv('path_to_table.csv') }}
```
Or an Excel like:
```
{{ read_excel('path_to_table.xlsx') }}
```
This plugin uses the [Pandas](https://pandas.pydata.org/) library, so this function should comply with that syntax.

No need to roll your own Super Fences or custom Java Script processes.

**Reply by roaldarbol** [reply]
That's really amazing - I'll revisit this then, since I never got on from from it and kind of stalled. Thanks!

**Reply by chrimaho** [reply]
Also related:
- #5080
- #5081

**Reply by roaldarbol** [reply]
So just to check, as it uses Pandas, is filtering possible - something like:
```
{{ read_csv('path_to_table.csv').query('population < 10000') }}
```

**Reply by timvink** [reply]
Hi, author of `mkdocs-table-reader-plugin` here. Filtering using `.query()` is definitely possible, but not arbitrary. The plugin will not just run any python inside `{{ }}` because that is potentially unsafe. Instead, I extract the function and arguments using regex and use `ast.literal_eval()`. 

I'm not sure if it's a good design choice though. You could use a hook to preprocess tables also (see [preprocessing tables using hooks](https://timvink.github.io/mkdocs-table-reader-plugin/howto/preprocess_tables/)), which is probably a better design pattern (because then you don't have any data logic inside your documentation code, it's just input/output).

I've created [an issue on the repo](https://github.com/timvink/mkdocs-table-reader-plugin/issues/44), let me know if you still find this useful compared to using a hook.

Thanks for pointing me to https://github.com/coddingtonbear/obsidian-csv-table, I like that syntax as well. I've already created another plugin using custom fences (https://github.com/timvink/mkdocs-charts-plugin) and I can confirm that it's really easy to do. I've created another issue for that one, if there are some upvotes I'll implement it: https://github.com/timvink/mkdocs-table-reader-plugin/issues/45
